### PR TITLE
Graph layout

### PR DESCRIPTION
**Pull request by karliss**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://cutter.re/docs/developers-docs/first-time.html) to this repository
- [ ] I made sure to follow the project's [coding style](https://cutter.re/docs/developers-docs.html) **Needs cleanup**
- [x] I've updated the [documentation](https://cutter.re/docs/user-docs.html) with the relevant information (if needed)

**Detailed description**

Rewrite the edge routing and block placement parts of graph layout. Add documentation for the graph layout algorithm.

* Take full subtree shape into account when placing them side by side. Prevents the case where 1-2  basic blocks are put very far to the right and then the edge goes back to main part of function.
long_edges
* In the basic true/false branching case horizontal lines are aligned at the same y level
* Heuristics for reducing edge crossing near the basic block
edge_horizontal_crossing

* Gets rid of unnecessary big  loops due to large basic block in the same row or column
* In single column case prefer routing false branches on one side and true branches on other side.
edge_loops_side

* Improved worst case CPU and memory usage. Some of the tables in old code could have been O(n^2) in the worst case. This wasn't a big problem with old code, as it was good enough in most cases with typical size functions and querying disassembly data from r2 takes an order of magnitude   longer than graph layout. It could be beneficial in future for drawing other r2 graphs like whole program xref graph.

**Test plan (required)**

Look for layout bugs most importantly overlapping blocks or edges
* Sort functions by size and look at biggest ones
* Look at small and medium size functions. 

**Feedback wanted**
I'd like some feedback for default edge and block spacing settings.

I have noticed that sometimes the layout for the same function gets draw in two different ways. I am not sure if lack of determinism comes from the order of nodes in r2 output or the graph layout process. If anyone can reliably reproduce this let me know.

**Closing issues**

**Follow up tasks**
Expose some of the layout tweaking settings in Perfrences/Graph

### CODE REVIEW

File: src/widgets/GraphGridLayout.cpp [hunk 1]
```diff
@@ -6,17 +6,142 @@
...
+following them would require going in both directions of edges. Example 1->2 1->3 3->2 is a DAG, 2->1 1->3 3->2
+isn't a DAG.
+- **DFS** - depth first search, a graph traversal algorithm
+- **toposort** - toplogical sorting, process of ordering a DAG vertices that all edges go from vertices erlier in the
+toposort order to vertices later in toposort order. There are multiple algorithms for implementing toposort operation.
+Single DAG can have multiple valid topoligical orderings, a toposort algorithm can be designed to priotarize a specific
+one from all valid toposort orders. Example: for graph 1->4, 2->1, 2->3, 3->4 valid topological orders are [2,1,3,4] and
+[2,3,1,4].
+
+# High level strucutre of the algorithm
```
Comment (xarkes) [inline src/widgets/GraphGridLayout.cpp:30]: structure

File: src/widgets/GraphGridLayout.cpp [hunk 2]
```diff
@@ -6,17 +6,142 @@
...
+6. calculate column and row pixel positions based on node sizes and amount edges between the rows
+
+
+Contrary to many other layered graph drawing algorithm this implementation doesn't perform node reording to minimize
+edge crossing. This simplifies implementation, and preserves original control flow structure for conditional jumps (
+true jump on one side, false jump on other). Due to most of control flow being result of structured programming
+constructs like if/then/else and loops, resulting layout is usually readable without node reordering within layers.
+
+
+# Describtion of grid.
```
Comment (xarkes) [inline src/widgets/GraphGridLayout.cpp:45]: Description

File: src/widgets/GraphGridLayout.h [hunk 3]
```diff
@@ -20,72 +25,147 @@ class GraphGridLayout : public GraphLayout
...
-    GridEdge routeEdge(EdgesVector &horiz_edges, EdgesVector &vert_edges,
-                       Matrix<bool> &edge_valid, GridBlock &start, GridBlock &end) const;
-    static int findVertEdgeIndex(EdgesVector &edges, int col, int min_row, int max_row);
-    static bool isEdgeMarked(EdgesVector &edges, int row, int col, int index);
-    static void markEdge(EdgesVector &edges, int row, int col, int index, bool used = true);
-    static int findHorizEdgeIndex(EdgesVector &edges, int row, int min_col, int max_col);
+    /**
+     * @brief Assign row positions to nodes.
+     * @param state
+     * @param blockOrder reverse toplogical ordering of nodes
```
Comment (xarkes) [inline src/widgets/GraphGridLayout.h:121]: topological

File: src/widgets/GraphGridLayout.h [hunk 4]
```diff
@@ -20,72 +25,147 @@ class GraphGridLayout : public GraphLayout
...
+     */
+    static void assignRows(LayoutState &state, const std::vector<ut64> &blockOrder);
+    /**
+     * @brief Select subset of DAG edges that form tree.
+     * @param state
+     * @param blockOrder reverse toplogical ordering of nodes
```
Comment (xarkes) [inline src/widgets/GraphGridLayout.h:127]: topological

File: src/widgets/GraphGridLayout.h [hunk 5]
```diff
@@ -20,72 +25,147 @@ class GraphGridLayout : public GraphLayout
...
+    /**
+     * @brief Do rough edge routing within grid using up to 5 segments.
+     */
+    void roughRouting(LayoutState &state) const;
+    /**
+     * @brief Calculate segment placement relative to their columns.
+     */
+    void elaborateEdgePlacement(LayoutState &state) const;
+    /**
+     * @brief Recalculate column widths, trying to compenstate for the space taken by edge columns.
```
Comment (xarkes) [inline src/widgets/GraphGridLayout.h:148]: compensate?

File: src/widgets/GraphGridLayout.cpp [hunk 6]
```diff
@@ -6,17 +6,142 @@
...
+Rows are assigned in toposort order from top to bottom, with nodes row being max(predecessor.row)+1. This ensures
+that loop edges are only ones going from deeper levels to previous layers.
+
+To further simply node placment a subset of edges is selected which forms a tree. This turns DAG drawing problem
+into a tree drawing problem. For each node in level n following nodes which have level exactly n+1 are greedily
+assigned as child nodes in tree. If a node already has perant assigned then corresponding edge is not part of tree.
+
+# Node position assignment
+
+Since the graph has been reduced to a tree node placement is more or less putting subtrees side by side with
```
Comment (xarkes) [inline src/widgets/GraphGridLayout.cpp:75]: `to a tree, node placement is...`? (added comma?)

File: src/widgets/GraphGridLayout.cpp [hunk 7]
```diff
@@ -103,438 +236,841 @@ void GraphGridLayout::CalculateLayout(std::unordered_map<ut64, GraphBlock> &bloc
...
 void GraphGridLayout::computeAllBlockPlacement(const std::vector<ut64> &blockOrder,
                                                LayoutState &layoutState) const
 {
+    assignRows(layoutState, blockOrder);
+    selectTree(layoutState);
+    findMergePoints(layoutState);
+
+
+    // Shapes of subtrees are maintained using linked lists. Each value within list is column relative to previous row.
+    // This allows moving things around by chaning only first value in list.
```
Comment (xarkes) [inline src/widgets/GraphGridLayout.cpp:341]: changing?

**Comment by ITAYC0HEN** [comment]
I said it before, and I will continue saying this -
This pull request is amazing. The improvements are very significant.


I didn't notice any bugs with the binaries, the symetricity is good, and the spacing is fine. Something It feels too narrow, and sometimes too wide. But I guess having it perfect is hard.

Will go over the code now, but I don't think I have the qualities to preform decent Code Review here. I asked @xarkes and @thestr4ng3r to do so - they have a better CS background than me with all this algorithms things. <3

cheers!

**Comment by thestr4ng3r** [comment]
Do you think you can make the commit history a bit nicer for rebase+merge or should we squash it?

**Comment by karliss** [comment]
I was thinking of squashing it.

**Pull request merged**